### DISCUSSION TITLE
Don't know how to do a delayed execution

### DISCUSSION THREAD

**Discussion by spiderbit** [stumpwm/stumpwm#1114] [Q&A]
basically I want to start as example firefox and then after it's fully loaded send a character to it (something I have vimium bound to).
```cl
(defcommand ff-start () ()
	    (run-shell-command "firefox")
	    (run-with-timer 3 nil	  
			    (run-commands "window-send-string u")))
```

I thought run-with-timer is for that but it just crashes stumpwm, also (sleep 3) or "firefox; sleep 3" doesn't work it just executes all parallel without run-with-timer and crashes it otherwise.

Do I misinterpret what run-with-timer should do or do I misuse it? Or is there another way to do what I want?

**Comment by jorams** [comment]
You're calling `run-with-timer` incorrectly. Something like this should work:

```lisp
(defcommand ff-start () ()
  (run-shell-command "firefox")
  (run-with-timer 3 nil #'window-send-string "u"))
```

**Comment by spiderbit** [comment]
@jorams It seems to work kind of, except that it runs when I reload instead of when press the button I have configured for it. any Idea why that is?

Well that happens if I define the keymap with setq or setf but with defvar it doesn't update my code with reload.

```cl
(defvar *e-map*
      (let ((m (make-sparse-keymap)))
	(define-key m (kbd "u") (ff-start))
	m))
```

**Comment by jorams** [comment]
@spiderbit As long as you don't call `ff-start`, it does not run. You probably have a call to ff-start somewhere in your config.

`defvar` does not assign the provided value if the variable already has a value. `defparameter` unconditionally assigns the value, so you probably want that one. See the [HyperSpec](http://www.lispworks.com/documentation/HyperSpec/Body/m_defpar.htm).

**Comment by spiderbit** [comment]
Ahh thx had with some errors crashes so I try it later and report back.
